lib/charms/observability_libs/v1/cert_handler.py

```diff
@@ -0,0 +1,391 @@
+
+    on = CertHandlerEvents()  # pyright: ignore
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        *,
+        key: str,
+        certificates_relation_name: str = "certificates",
+        cert_subject: Optional[str] = None,
+        sans: Optional[List[str]] = None,
+    ):
+        """CertHandler is used to wrap TLS Certificates management operations for charms.
+
+        CerHandler manages one single cert.
+
+        Args:
+            charm: The owning charm.
+            key: A manually-crafted, static, unique identifier used by ops to identify events.
+             It shouldn't change between one event to another.
+            certificates_relation_name: Must match metadata.yaml.
+            cert_subject: Custom subject. Name collisions are under the caller's responsibility.
+            sans: DNS names. If none are given, use FQDN.
+        """
+        super().__init__(charm, key)
+
+        self.charm = charm
+        # We need to sanitize the unit name, otherwise route53 complains:
+        # "urn:ietf:params:acme:error:malformed" :: Domain name contains an invalid character
+        self.cert_subject = charm.unit.name.replace("/", "-") if not cert_subject else cert_subject
+
+        # Use fqdn only if no SANs were given, and drop empty/duplicate SANs
+        sans = list(set(filter(None, (sans or [socket.getfqdn()]))))
+        self.sans_ip = list(filter(is_ip_address, sans))
+        self.sans_dns = list(filterfalse(is_ip_address, sans))
+
+        self.certificates_relation_name = certificates_relation_name
+        self.certificates = TLSCertificatesRequiresV2(self.charm, self.certificates_relation_name)
+
+        self.framework.observe(
+            self.charm.on.config_changed,
+            self._on_config_changed,
+        )
+        self.framework.observe(
+            self.charm.on.certificates_relation_joined,  # pyright: ignore
+            self._on_certificates_relation_joined,
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_available,  # pyright: ignore
+            self._on_certificate_available,
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_expiring,  # pyright: ignore
+            self._on_certificate_expiring,
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_invalidated,  # pyright: ignore
+            self._on_certificate_invalidated,
+        )
+        self.framework.observe(
+            self.certificates.on.all_certificates_invalidated,  # pyright: ignore
+            self._on_all_certificates_invalidated,
+        )
+        self.framework.observe(
+            self.charm.on[self.certificates_relation_name].relation_broken,  # pyright: ignore
+            self._on_certificates_relation_broken,
+        )
+
+    @property
+    def enabled(self) -> bool:
+        """Boolean indicating whether the charm has a tls_certificates relation."""
```

Comment:
... and that the relation is healthy (there is a remote app, there is a databag...) to filter out some common 'relation-breaking' data model inconsistencies.